### DISCUSSION TITLE
Details of defer/stream implementation

### DISCUSSION THREAD

**Discussion by mattfysh** [graphql/graphql-js#3985] [General]
Hi all, thanks for all the recent work on the defer/stream implementation. I've got some ideas I'm excited to prototype!

Looking through the source code now, and I had a few questions coming up that aren't so straightforward to answer from reading the code. Are there any design documents that explain things like "target" and "incremental data record", how they are related & which features they support?

A couple of unrelated other questions
- the code became much easier to grok once I realized `exeContext.incrementalPublisher` is a one-way communication channel. Would there be any benefit in having something in the execution engine that is an event emitter and then the incremental publisher can process the events?
- many lines of code in execute.ts are used only for prop drilling, I wonder if there's value in using an `..args` spread to reduce verbosity?
- another thing that may help is having more utilities to handle the common pattern: `try { if (promise) { X().catch(Y) } else { X() }  } catch(e) { Y() }`
- lastly, has there been much discussion on having a `@live` directive which could benefit from the incremental publisher?

Thanks again 🙌  you guys 🪨

**Comment by robrichard** [comment]
@mattfysh the defer/stream spec proposal is under active development and the mechanics of how it works has changed recently (see https://github.com/graphql/defer-stream-wg/discussions/69). `graphql-js` now contains an updated implementation that we hope can be used to gather feedback around the new response format. However, we are still very much working out the details of how this will be specified in the GraphQL specification. My hope is that once these details are finalized, the spec will explain all of the concepts necessary to understand and implement defer & stream and the implementation in `graphql-js` should closely match what is specified. 

I know that @IvanGoncharov has been looking into ways to clean up the code, especially around return values that may or may not be promises.

I am not aware of any recent discussions around `@live` at any GraphQL working group meetings
